### PR TITLE
Improves RTL style conversion

### PR DESCRIPTION
**Pull request by ItsJonQ**
## Description
This update improves the rtl style utility to better handle varying combinations of `left` and `Left` based properties. A handful of tests have been added to ensure style property conversions are outputted as expected.

## Types of changes
* Adjusted the regex logic for `left` and `Left` replacements in the `rtl` convertor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. 

Addresses (🤞) @aduth 's comments from https://github.com/WordPress/gutenberg/pull/19916/files/110831f8f61f5e43ff9c76331159415040b412c9#r379631860

### CODE REVIEW

**Review by aduth — CHANGES_REQUESTED** [review]
I didn't remark on it originally, but shouldn't we be switching instances of "right" to "left" as well?

**Comment by ItsJonQ** [comment]
@aduth Ah yes! Thank you so much <3. I really appreciate your thoroughness and feedback! I guess I got a little tunnel-visioned on just the left -> right aspect of it.

I just made an update and added tests for left -> right, and right -> left.

**Comment by aduth** [comment]
Now I can add the edge case test I had planned to add yesterday 😄 See. 474455450 (currently failing)

File: packages/components/src/utils/test/rtl.test.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
+/**
```
Comment (aduth) [inline packages/components/src/utils/test/rtl.test.js]: We've a lot of inconsistency in the project, but the project guidelines are such that I would expect this file should be named `test/rtl.js`.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/testing-overview.md#folder-structure

Comment (ItsJonQ) [reply packages/components/src/utils/test/rtl.test.js]: Ah, gotcha! Will make that update! 🤗

**Comment by ItsJonQ** [comment]
@aduth TIL about `text-combine-upright` 😱

**Comment by aduth** [comment]
@ItsJonQ I had never heard of it previously either 😄 Just pulled it as an example looking through [a full list of CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference) for substrings of "left" and "right" which we'd not expect to flip in the sense of horizontal directionality.

File: packages/components/src/utils/rtl.js [hunk 2]
```diff
@@ -20,18 +25,35 @@ export function useRtl() {
...
+		if ( key === 'left' ) {
+			nextKey = 'right';
+		}
+		if ( key === 'right' ) {
+			nextKey = 'left';
+		}
+		// Lowercase flip
+		if ( lowerLeftRegExp.test( key ) ) {
+			nextKey = key.replace( lowerLeftRegExp, '-right' );
 		}
```
Comment (aduth) [inline packages/components/src/utils/rtl.js]: Per [my reference comment](https://github.com/WordPress/gutenberg/pull/19916/files/110831f8f61f5e43ff9c76331159415040b412c9#r379631860), would there be any difference if this was implemented as:

```suggestion
		nextKey = key.replace( lowerLeftRegExp, '-right' );
```

Comment (ItsJonQ) [reply packages/components/src/utils/rtl.js]: @aduth I just tried. It breaks one of the conversions in the unit tests

```
  convertLtrToRtl
    ✓ converts (*)Left <-> (*)Right (6ms)
    ✕ converts (*)left <-> (*)right (7ms)

  ● convertLtrToRtl › converts (*)left <-> (*)right

    expect(received).toBe(expected) // Object.is equality

    Expected: 10
    Received: 20

      105 | 		expect( nextStyle[ 'scroll-margin-right' ] ).toBe( 10 );
      106 | 		expect( nextStyle[ 'scroll-padding-right' ] ).toBe( 10 );
    > 107 | 		expect( nextStyle.right ).toBe( 10 );
          | 		                          ^
      108 |
      109 | 		// right -> left
      110 | 		expect( nextStyle[ 'border-left' ] ).toBe( '20px solid blue' );

      at Object.toBe (packages/components/src/utils/test/rtl.js:107:29)
```

File: packages/components/src/utils/rtl.js [hunk 3]
```diff
@@ -3,6 +3,11 @@
  */
 import { css } from '@emotion/core';
 
+const lowerLeftRegExp = new RegExp( /-left/g );
+const lowerRightRegExp = new RegExp( /-right/g );
+const upperLeftRegExp = new RegExp( /Left/g );
+const upperRightRegExp = new RegExp( /Right/g );
```
Comment (aduth) [inline packages/components/src/utils/rtl.js]: They look like constants to me.

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#constants

```suggestion
const LOWER_LEFT_REGEXP = new RegExp( /-left/g );
const LOWER_RIGHT_REGEXP = new RegExp( /-right/g );
const UPPER_LEFT_REGEXP = new RegExp( /Left/g );
const UPPER_RIGHT_REGEXP = new RegExp( /Right/g );
```

File: packages/components/src/utils/rtl.js (hunk 2 above)
Comment (aduth) [reply packages/components/src/utils/rtl.js]: I think the reason is that we should be doing the replace as operating on the working copy of `nextKey`.

```suggestion
		nextKey = nextKey.replace( lowerLeftRegExp, '-right' );
```

**Comment by aduth** [comment]
More a minor point, but could become a scaling issue: We should be able to stop immediately once we reach a match of any of these conditions, rather than continue to test them. It would probably be a refactor to a separate function using `return`:

```js
function getConvertedKey( key ) {
	if ( key === 'left' ) {
		return 'right';
	}
	
	if ( key === 'right' ) {
		return 'left';
	}

	return key
		.replace( LOWER_LEFT_REGEXP, '-right' )
		.replace( LOWER_RIGHT_REGEXP, '-left' )
		.replace( UPPER_LEFT_REGEXP, '-right' )
		.replace( UPPER_RIGHT_REGEXP, '-left' );
}
```

Though now that I write this out, it only really benefits the direct `'right'` and `'left'`, unless we keep the `RegExp#test`.

**Comment by aduth** [comment]
Actually, in testing that, I see now the problem with always running `replace` on the working copy of the key (rather than the original) is that a later replace will undo the effects of an earlier.

So a correct optimized version would be:

```js
function getConvertedKey( key ) {
	if ( key === 'left' ) {
		return 'right';
	}
	
	if ( key === 'right' ) {
		return 'left';
	}

    if ( LOWER_LEFT_REGEXP.test( key ) ) {
        return key.replace( LOWER_LEFT_REGEXP, '-right' );
    }

    if ( LOWER_RIGHT_REGEXP.test( key ) ) {
        return key.replace( LOWER_RIGHT_REGEXP, '-left' );
    }

    if ( UPPER_LEFT_REGEXP.test( key ) ) {
        return key.replace( UPPER_LEFT_REGEXP, 'Right' );
    }

    if ( UPPER_RIGHT_REGEXP.test( key ) ) {
        return key.replace( UPPER_RIGHT_REGEXP, 'Left' );
    }

    return key;
}
```

**Comment by ItsJonQ** [comment]
> running replace on the working copy of the key (rather than the original) is that a later replace will undo the effects of an earlier.

@aduth That's what I ran into as well 😂 . The chained replace would work if there were unique temporary tokens injected, which would be converted to desired values (at the very end).

Your refactor suggestion is great! I'll make the update

**Review by aduth — APPROVED** [review]

**Comment by aduth** [comment]
Another possible simplification:

```diff
diff --git a/packages/components/src/utils/rtl.js b/packages/components/src/utils/rtl.js
index 317d8eb4a..cf366412e 100644
--- a/packages/components/src/utils/rtl.js
+++ b/packages/components/src/utils/rtl.js
@@ -2,6 +2,7 @@
  * External dependencies
  */
 import { css } from '@emotion/core';
+import { mapKeys } from 'lodash';
 
 const LOWER_LEFT_REGEXP = new RegExp( /-left/g );
 const LOWER_RIGHT_REGEXP = new RegExp( /-right/g );
@@ -68,17 +69,8 @@ function getConvertedKey( key ) {
  *
  * @return {Object} Converted ltr -> rtl styles
  */
-export const convertLtrToRtl = ( ltrStyles = {} ) => {
-	const nextStyles = {};
-
-	for ( const key in ltrStyles ) {
-		const value = ltrStyles[ key ];
-		const nextKey = getConvertedKey( key );
-		nextStyles[ nextKey ] = value;
-	}
-
-	return nextStyles;
-};
+export const convertLtrToRtl = ( ltrStyles ) =>
+	mapKeys( ltrStyles, ( _value, key ) => getConvertedKey( key ) );
 
 /**
  * A higher-order function that create an incredibly basic ltr -> rtl style converter for CSS objects.
```

File: packages/components/src/utils/rtl.js [hunk 4]
```diff
@@ -3,47 +3,90 @@
...
+
 /**
  * An incredibly basic ltr -> rtl converter for style properties
  *
  * @param {Object} ltrStyles
+ *
  * @return {Object} Converted ltr -> rtl styles
  */
-const convertLtrToRtl = ( ltrStyles = {} ) => {
+export const convertLtrToRtl = ( ltrStyles = {} ) => {
```
Comment (aduth) [inline packages/components/src/utils/rtl.js]: On the topic of case standards, there's also:

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#abbreviations-and-acronyms

i.e. This should technically be `convertLTRToRTL`

Comment (ItsJonQ) [reply packages/components/src/utils/rtl.js]: Thank you!! Interesting question... In that case, should the main function be renamed from `rtl()` to `RTL()`?

Comment (aduth) [reply packages/components/src/utils/rtl.js]: `rtl` would be the correct capitalization. The phrasing is a bit strange, but it's meant to be covered by this bit of the standard:

>If an abbreviation or an acronym occurs at the start of a variable name, it must be written to respect the camelcase naming rules covering the first letter of a variable or class definition. For variable assignment, this means writing the abbreviation entirely as lowercase.

**Comment by ItsJonQ** [comment]
Hmm.. not sure how to re-run Travis. I clicked "Re-run" on this page (https://github.com/WordPress/gutenberg/pull/20503/checks?check_run_id=475805503), but it didn't seem to do anything

**Comment by aduth** [comment]
@ItsJonQ I've been encountering the same in my pull request at #20544. Not sure yet if it might be an outage from where we download Chromium (for Puppeteer / E2E tests).

**Pull request merged**